src/azure-cli-core/azure/cli/core/commands/progress.py

```diff
@@ -0,0 +1,181 @@
+    """ generic progress reporter """
+    def __init__(self):
+        self.message = ''
+
+    def add(self, message=''):
+        """ adds a progress report """
+        self.message = message
+
+    def report(self):
+        """ report the progress """
+        return self.message
+
+
+class ProgressHook(object):
+    """ sends the progress to the view """
+    def __init__(self, progress_type):
+        if progress_type == ProgressType.Determinate:
+            self.reporter = DetProgressReporter()
+        elif progress_type == ProgressType.Indeterminate:
+            self.reporter = InDetProgressReporter()
+        self.progress_type = progress_type
+        from azclishell.progress import ShellProgressView
+        self.registery = [IndeterminateStandardOut, DeterminateStandardOut, ShellProgressView]
+        self.active_progress = []
+
+    def init_progress(self, progress_view):
+        """ activate a view """
+        assert any(isinstance(progress_view, view) for view in self.registery)
+        self.active_progress.append(progress_view)
+
+    def register_view(self, progress_view):
+        """ register a view to report to """
+        assert issubclass(progress_view, _ProgressViewBase)
+        self.registery.append(progress_view)
+
+    def add(self, **kwargs):
+        """ adds a progress report """
+        self.reporter.add(kwargs)
+        self.update()
+
+    def update(self):
+        """ updates the view with the progress """
+        for view in self.active_progress:
+            view.write(self.reporter.report())
+            view.flush()
+
+    def begin(self):
+        """ start reporting progress """
+        if self.progress_type == ProgressType.Determinate:
+            self.add(message='Starting', value=0.0, total_val=1)
+        else:
+            self.add(message='Starting')
+
+    def end(self):
+        """ ending reporting of progress """
+        if self.progress_type == ProgressType.Determinate:
+            self.add(message="Finished", value=1.0, total_val=1)
+        else:
+            self.add(message='Finished')
+
+
+class IndeterminateStandardOut(InDeterminateProgressView):
+    """ custom output for progress reporting """
+    def __init__(self, out=None):
+        super(IndeterminateStandardOut, self).__init__(out if out else sys.stderr)
+        self.spinner = humanfriendly.Spinner(label='In Progress')
+        self.spinner.hide_cursor = False
+
+    def write(self, args):
+        """ writes the progress """
+        msg = args['message'] if 'message' in args else 'In Progress'
+        self.spinner.step(label=msg)
+
+
+class DeterminateStandardOut(DeterminateProgressView):
+    """ custom output for progress reporting """
+    def __init__(self, out=None):
+        super(DeterminateStandardOut, self).__init__(
+            out if out else sys.stderr, self._format_value)
+
+    def _format_value(self, msg, percent):  # pylint: disable=no-self-use
+        bar_len = BAR_LEN - len(msg) - 1
```

Comment:
I assume this doesn't work if `bar_len < 0`?
To prevent `bar_len` from going negative, you could truncate `msg` if it's too long. (see `truncate_text() util method in azure.cli.core.util`)